lib/contracts/portal/deploy.go

```diff
@@ -0,0 +1,208 @@
+package portal
+
+import (
+	"context"
+
+	"github.com/omni-network/omni/contracts/bindings"
+	"github.com/omni-network/omni/lib/chainids"
+	"github.com/omni-network/omni/lib/contracts"
+	"github.com/omni-network/omni/lib/contracts/create3"
+	"github.com/omni-network/omni/lib/errors"
+	"github.com/omni-network/omni/lib/ethclient/ethbackend"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	ethtypes "github.com/ethereum/go-ethereum/core/types"
+)
+
+type DeploymentConfig struct {
+	Create3Factory common.Address
+	Create3Salt    string
+	ProxyAdmin     common.Address
+	Deployer       common.Address
+	Owner          common.Address
+	FeeOracle      common.Address
+	ValSetID       uint64
+	Validators     []bindings.Validator
+	ExpectedAddr   common.Address
+}
+
+func (cfg DeploymentConfig) Validate() error {
+	if (cfg.Create3Factory == common.Address{}) {
+		return errors.New("create3 factory is zero")
+	}
+	if cfg.Create3Salt == "" {
+		return errors.New("create3 salt is empty")
+	}
+	if (cfg.ProxyAdmin == common.Address{}) {
+		return errors.New("proxy admin is zero")
+	}
+	if (cfg.Deployer == common.Address{}) {
+		return errors.New("deployer is zero")
+	}
+	if (cfg.Owner == common.Address{}) {
+		return errors.New("owner is zero")
+	}
+	if (cfg.FeeOracle == common.Address{}) {
+		return errors.New("fee oracle is zero")
+	}
+	if cfg.ValSetID == 0 {
+		return errors.New("validator set ID is zero")
+	}
+	if len(cfg.Validators) == 0 {
+		return errors.New("validators is empty")
+	}
+
+	return nil
+}
+
+func getDeployCfg(chainID uint64, network string) (DeploymentConfig, error) {
+	if chainids.IsMainnet(chainID) {
+		return mainnetDeployCfg(network), nil
+	}
+
+	if chainids.IsTestnet(chainID) {
+		return testnetDeployCfg(network), nil
+	}
+
+	return DeploymentConfig{}, errors.New("unsupported chain id")
+}
+
+func mainnetDeployCfg(network string) DeploymentConfig {
+	return DeploymentConfig{
+		Create3Factory: contracts.MainnetCreate3Factory,
+		Create3Salt:    network + "-portal",
```

Comment:
Is network (a) `mainnet` `testnet` etc or (b) `optimism` `arbitrum` etc

if (a) - could we just have a `deployCfg(network string)` that returns based on the string passed in? rather than `mainnetDeployCfg("mainnet")`

if (b) - I don't think that works, since it will create a different portal address on each chain